### PR TITLE
Add SignInButton with G+ style

### PR DESCRIPTION
**Pull request by tomrozb**

### CODE REVIEW

**Comment by tomrozb** [comment]
I'm going to add some tests in a new commit. It's definitely not TDD :)

File: lib-android-common/src/main/java/com/devinci/lib/util/Versions.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.devinci.lib.util;
+
+import android.os.Build;
+
+/**
+ * Utility methods to check Android version.
+ */
+public enum Versions {
+  ;
```
Comment (arturdm) [inline lib-android-common/src/main/java/com/devinci/lib/util/Versions.java]: I have mixed feelings with this util class approach - auto-completion suggests enum methods :/

File: lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import com.devinci.lib.util.Versions;
+
+public enum ViewUtils {
+  ;
+
+  @TargetApi(Build.VERSION_CODES.LOLLIPOP) @SuppressWarnings("deprecation")
+  public static Drawable getDrawable(Context context, @DrawableRes int drawable) {
+    Resources resources = context.getResources();
+    if (Versions.isLollipop()) {
+      return resources.getDrawable(drawable, context.getTheme());
```
Comment (arturdm) [inline lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java]: You can use compat version:

``` java
ResourcesCompat.getDrawable(context.getResources(), drawable, context.getTheme());
```

File: lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    } else {
+      return resources.getDrawable(drawable);
+    }
+  }
+
+  @TargetApi(Build.VERSION_CODES.M) @SuppressWarnings("deprecation")
+  public static int getColor(Context context, @ColorRes int color) {
+    Resources resources = context.getResources();
+    if (Versions.isMarshmallow()) {
+      return resources.getColor(color, context.getTheme());
```
Comment (arturdm) [inline lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java]: There's `ContextCompat.getColor(context, color);`.

File: lib-android/src/main/res/layout/lib_button_sign_in.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    xmlns:tools="http://schemas.android.com/tools"
+    tools:background="@color/lib_sign_in_google"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:layout_height="50dp"
+    >
+
+  <ImageView
+      android:id="@+id/lib_button_sign_in_icon"
+      android:layout_width="48dp"
```
Comment (arturdm) [inline lib-android/src/main/res/layout/lib_button_sign_in.xml]: Shouldn't those values be moved to dimens too?

File: lib-android/src/main/res/layout/lib_button_sign_in.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      />
+
+  <TextView
+      android:id="@+id/lib_button_sign_in_text"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_centerInParent="true"
+      android:gravity="center"
+      android:textColor="@color/lib_white"
+      android:textSize="15sp"
```
Comment (arturdm) [inline lib-android/src/main/res/layout/lib_button_sign_in.xml]: Shouldn't be styled?

File: lib-android-common/src/main/java/com/devinci/lib/util/Versions.java (hunk 1 above)
Comment (tomrozb) [inline lib-android-common/src/main/java/com/devinci/lib/util/Versions.java]: Ok, so we're not going to use this anymore. I'll switch to private constructor.

File: lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    SignInButton signInButton = new SignInButton(context);
+    Drawable expectedIcon = ViewUtils.getDrawable(context, R.drawable.lib_ic_sign_in_google);
+
+    signInButton.setIcon(R.drawable.lib_ic_sign_in_google);
+
+    assertThat(signInButton.getIcon()).isEqualTo(expectedIcon);
+  }
+
+  @Test public void shouldSetText() throws Exception {
+    SignInButton signInButton = new SignInButton(context);
```
Comment (arturdm) [inline lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java:94]: I see that you use newlines to separate blocks. We can try that.

File: lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    assertThat(signInButton.getForeground()).isEqualTo(expectedForeground);
+    assertThat(signInButton.getBackground()).isEqualTo(expectedBackground);
+    assertThat(signInButton.getText()).isEqualTo(expectedTitle);
+    assertThat(signInButton.getIcon()).isEqualTo(expectedIcon);
+  }
+
+  @Test public void shouldSetIcon() throws Exception {
+    SignInButton signInButton = new SignInButton(context);
+    Drawable expectedIcon = ViewUtils.getDrawable(context, R.drawable.lib_ic_sign_in_google);
```
Comment (arturdm) [inline lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java]: You can compare resource id by using `shadowOf(signInButton.getIcon()).getCreatedFromResId()`.

Comment (tomrozb) [inline lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java]: +1

File: lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java (hunk 6 above)
Comment (tomrozb) [inline lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java:94]: We can mix styles, both are easy to read

File: lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java (hunk 2 above)
Comment (tomrozb) [inline lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java]: Should I remove ViewUtils and use ResourcesCompat or just replace ViewUtils implementation?

File: lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java (hunk 3 above)
Comment (tomrozb) [inline lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java]: +1

File: lib-android/src/main/res/layout/lib_button_sign_in.xml (hunk 4 above)
Comment (tomrozb) [inline lib-android/src/main/res/layout/lib_button_sign_in.xml]: I'm going to move most of the values to styles or attrs, so the SignInButton can be easily styled.

File: lib-android/src/main/res/layout/lib_button_sign_in.xml (hunk 5 above)
Comment (tomrozb) [inline lib-android/src/main/res/layout/lib_button_sign_in.xml]: +1

File: lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java (hunk 2 above)
Comment (arturdm) [inline lib-android/src/main/java/com/devinci/lib/view/ViewUtils.java]: I'm not sure if compat classes APIs change a lot, but I guess we can use them directly and if APIs brake we can think about wrapping them in our classes.

File: lib-android/src/main/res/values/colors.xml [hunk 8]
```diff
@@ -4,4 +4,9 @@
   <color name="lib_transparent">#00000000</color>
   <color name="lib_white">#FFFFFFFF</color>
   <color name="lib_black">#FF000000</color>
+
+  <!-- SignInButton -->
+  <color name="lib_sign_in_google">#FFE73F25</color>
```
Comment (arturdm) [inline lib-android/src/main/res/values/colors.xml]: To discuss - what do you think about moving these specific values to other resource file with _singInButton suffix just like with the style? (applies to dimens and strings too)

File: lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricGradleTestRunner.class) public class SignInButtonTest {
+
+  private Context context;
+  private Drawable expectedBackground;
+  private Drawable expectedForeground;
+
+  @Before public void setUp() throws Exception {
```
Comment (arturdm) [inline lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java]: Throws in those methods can be omitted - applies to `@Before`, `@After` and `@Test` methods (and added back if there really are exceptions that can be thrown).

File: lib-android/src/main/res/values/colors.xml (hunk 8 above)
Comment (tomrozb) [inline lib-android/src/main/res/values/colors.xml]: Moved

File: lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java (hunk 9 above)
Comment (tomrozb) [inline lib-android/src/test/java/com/devinci/lib/view/SignInButtonTest.java]: Done

File: lib-android/src/main/java/com/devinci/lib/view/SignInButton.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+  private void setTextInternal(String text) {
+    buttonText.setText(text);
+  }
+
+  @VisibleForTesting String getText() {
+    return buttonText.getText().toString();
+  }
+
+  public void setTextSize(@DimenRes int textSize) {
```
Comment (tomrozb) [inline lib-android/src/main/java/com/devinci/lib/view/SignInButton.java:106]: There's no way to assert text size. Robolectric always returns 14.0f as text size.

**Comment by arturdm** [comment]
+1

**Pull request merged**